src/fetch/PageAction.js

```diff
@@ -0,0 +1,113 @@
+import { pageAction } from "./prompts.js";
+import { logger } from "../log/logger.js";
+import { TagRemovingMinimizer } from "../min/TagRemovingMinimizer.js";
+import { getAI } from '../ai/index.js';
+import { Timer } from "../log/timer.js";
+
+export const PageAction = class {
+  constructor(namespace, prompt) {
+    this.namespace = namespace;
+    this.prompt = prompt;
+    this.ai = getAI()
+  }
+
+  async learn(examples) {
+
+  }
+
+  async execute(page) {
+    const timer = new Timer();
+    const url = page.url;
+    const doc = await page.content();
+    // const min = new TagRemovingMinimizer(['style', 'script', 'meta', 'link']);
+    // const minDoc = await min.min(doc, { timer });
+    const context = {
+      html: doc,
+      prompt: this.prompt
+    }
+
+    let prompts;
+    try {
+      prompts = await pageAction.renderMulti(context, 'html', this.ai);
+    } catch (e) {
+      logger.error(`${this} Error while rendering prompts: ${e}`);
+      return;
+    }
+
+    const commands = [];
+    logger.debug(`${this} analyze chunks for next action (${prompts.length})`);
+
+    for (const prompt of prompts) {
+      let answer;
+      try {
+        answer = await this.ai.ask(prompt, { format: 'json' });
+      } catch(e) {
+        logger.error(`${this} AI error, ignore: ${e}`);
+        continue
+      }
+
+      logger.debug(`${this} Got an answer: ${JSON.stringify(answer.partial)}`);
+
+      if (answer.partial?.actionCommand && answer.partial?.actionArgument) {
+        commands.push({
+          command: answer.partial.actionCommand,
+          arg: answer.partial.actionArgument,
+        });
+      }
+    }
+
+    let index = 0;
+    while (index < commands.length) {
+      const { command, arg } = commands[index];
+      logger.info(`${this} Taking action: ${command} ${arg}, index=${index}}`);
+      
+      try {
+        switch (command) {
+          case 'click':
+            await this.click(arg, page);
+            break;
+          case 'scroll':
+            await this.scroll(arg, page);
+            break;
+          default:
+            logger.error(`${this} Unhandled command: ${command} ${arg}`);
+            break;
+        }
+      } catch (e) {
+        logger.error(`${this} Error while executing action ${command} ${arg}, ignoring: ${e}`);
+      }
+      
+      index++;
```

Comment:
- the index++ in pagination was in the case where pagination failed
- here, we don't (right now) know if action succeeded or failed
- we could check if actions succeed or fail in learn(), or the first time we do the action in execute
- basically, capture a before/after snapshot, and then ask the AI if it seems like the action succeeded

this will likely work, but please test it a few times to make sure AI is giving good results